### DISCUSSION TITLE
Better protection against failing to protect tables with RLS

### DISCUSSION THREAD

**Discussion by BernieSumption** [supabase/supabase#11538] [Feature Requests]
I was recently surprised to discover that the default for tables that do not have RLS enabled is full read-write access to unauthenticated users.

When creating a table through the web UI there is a warning if you do not enable RLS, but when using SQL there is no warning, and tables have RLS disabled by default.

I struggle to think of any use case in a real publicly deployed application for intentionally exposing a public API to delete or overwrite all data in a table. Given how easy it is to do this accidentally, I would predict that the vast majority of your users' projects that expose their `anon public` key and have left RLS off on one or more tables have done so accidentally and have a serious security hole in their app.

Is some kind of safety net warranted? Potential mitigations include requiring RLS to be enabled in order to access a table through the API, or enabling RLS by default and requiring it to be explicitly disabled. Or if those go against the grain of the way that Postgres/PostgREST work, making it much more obvious in the UI. I'd have thought the existence of a table without RLS enabled in a project that uses the API was enough of a red flag to warrant a big red banner on every page in the UI, or emails sent to the project owner.

Firebase handles this quite well by allowing you to make all data publicly accessible for the first 30 days after creating your project, so that you don't need to learn about security rules on day 1, but can't leave your app unsecured indefinitely.

**Comment by steve-chavez** [comment]
> Or if those go against the grain of the way that Postgres/PostgREST work

Not really, PostgreSQL actually is strict on tables/views and you need to whitelist them to allow access. PostgREST goes a bit further and also recommends to have [functions access be a whitelist](https://postgrest.org/en/stable/schema_structure.html#functions)(the PostgreSQL default is a blacklist).

> I was recently surprised to discover that the default for tables that do not have RLS enabled is full read-write access to unauthenticated users.

There's a discussion on https://github.com/supabase/supabase/discussions/1493. I think with https://github.com/supabase/supabase/issues/11745 we may have a way forward.

> Firebase handles this quite well by allowing you to make all data publicly accessible for the first 30 days after creating your project, so that you don't need to learn about security rules on day 1, but can't leave your app unsecured indefinitely.

That's an interesting idea! So `anon` would have all its privileges revoked after 30 days.. will discuss this with the team internally.

**Reply by BernieSumption** [reply]
Thanks for the response.

A whitelist would definitely work for me. If there was a whitelist of tables and functions, having default read/write permissions for tables without RLS enabled wouldn't be an issue, since you'd taken the positive step of exposing the table without creating rules.

> That's an interesting idea! So anon would have all its privileges revoked after 30 days.. will discuss this with the team internally.

The firebase default rules for a project created today would be:

```
rules_version = '2';
service cloud.firestore {
  match /databases/{database}/documents {
    match /{document=**} {
      allow read, write: if
          request.time < timestamp.date(2023, 2, 19); // <- generated based on today's date
    }
  }
}
```

So there's nothing magic, they just use the rules syntax to define "allow everything until {date 30 days from project creation} then deny everything".

**Comment by BernieSumption** [comment]
I've discovered another surprising default - views are created with `security_invoker = false`, and also exposed in the API by default, so creating a view in the console is another way to expose read/write access to anonymous users.

Of course the bit about view default privileges is Postgres default behaviour, but for anyone coming from Firebase and learning Postgres security as part of the move, it's a bit of a footgun.
